### PR TITLE
MDEV-40924: validate insert_length before memcpy in KEY_OP_ADD_SUFFIX

### PR DESCRIPTION
**Pull request by shoemoney**
- [x] The Jira issue number for this PR is MDEV-40924.

## Description

KEY_OP_ADD_SUFFIX copied insert_length bytes into the key page without checking that the source range stayed inside the log record or that the result fit the page. A corrupt or truncated record could therefore read past header_end and write past max_page_size during redo.

Validate both bounds before the memcpy, and apply the matching check to KEY_OP_DEL_SUFFIX, which had the same shape. On failure, return an error and mark the file crashed through the existing error path. Keep the debug assertions.

The September 17 rebase onto 10.11 at `f4a3b69bdf14` preserves the original 12-line patch as one commit.

## How can this PR be tested?

The native Aria build and five CTests were rerun after this rebase on macOS arm64. The standalone recovery and API diagnostic below passed before the rebase; the Aria source is byte-identical.

- Native Aria libraries, tools and relevant test executables build successfully.
- Five relevant CTests pass with `MYTAP_CONFIG=big`, including the 8k pagecache assertions.
- The standalone recovery suite passes: `ALL RECOVERY TESTS OK (zerofilled 56 tables)`.
- A local diagnostic calls the actual `_ma_apply_redo_index` implementation with real pagecache and temporary file I/O. Eight valid, boundary and malformed-record cases pass with the patch. On current 10.11, the four malformed-record expectations fail while the four valid/boundary controls pass. Rejection checks cover the return value, crashed state in memory and on disk, and unchanged page contents.

The diagnostic is not yet integrated into the repository test suite; committed regression coverage remains outstanding. Its log buffers are physically padded while the logical record lengths are truncated. This verifies the bounds checks, but does not establish safety for every truncated allocation: the existing two-byte length reads still precede the new checks. Full debug MTR fault-injection coverage was not run.

## Basing the PR against the correct MariaDB version

- [x] This is a bug fix based on 10.11, as requested in preliminary review.

## Backward compatibility

No on-disk format changes. Valid suffix operations retain their behavior; the added checks reject invalid lengths before the copy or page-length update.

### CODE REVIEW

File: storage/maria/ma_key_recover.c [hunk 1]
```diff
@@ -1059,6 +1059,12 @@ uint _ma_apply_redo_index(MARIA_HA *info,
     {
       uint insert_length= uint2korr(header);
       DBUG_PRINT("redo", ("key_op_add_suffix: %u", insert_length));
+      if (header + 2 + insert_length > header_end ||
```
Comment (gkodinov) [inline storage/maria/ma_key_recover.c:1062]: Please consider trying to get a regression test for this. I know it's an edge condition, but we try to keep stuff tested as much as possible. Give it some effort please.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a preliminary review.

Please rebase on 10.11: this is the lowest affected version.
Please also attend to the CLAassistant check.

**Comment by shoemoney** [comment]
Rebased onto 10.11 as requested; PR now targets 10.11 with a single commit (5fee4e2). Still working on a regression test for the edge condition.

**Review by gkodinov — APPROVED** [review]
Looks great, thanks! While you are looking for the regression test, I'll procure a final reviewer.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Please also clear the CLA bot!

**Comment by shoemoney** [comment]
Rebased onto current 10.11 (10 commits, none touching ma_key_recover.c — clean, no diff changes). CLA bot is showing green now as well.

On the regression test: dug into it — `_ma_apply_redo_index` only runs from the real crash-recovery path, so there's no unit-test seam for it; it has to go through the mtr recovery harness (maria-recovery.test style: snapshot, crash, recover, diff). Planning to add a `DBUG_EXECUTE_IF` fault injector right after the `insert_length`/`del_length` reads (forcing an out-of-range value under a debug flag) and a new maria-recovery test that sets that flag via `.opt`, crashes, and asserts recovery marks the table crashed instead of dying. Will follow up with that once I can validate the `.result` output against a debug build rather than guessing at it blind.

**Review by gkodinov — APPROVED** [review]
LGTM. Thanks for considering a regression test. Please stand by for the final review.

**Comment by gkodinov** [comment]
FYI: According to our development cycle we work on bugs In the following periods 15 Mar-30 Apr, 15 Jun-30 Jul, 15 Sep-30 Oct and 15 Dec-31 Jan. So, please, expect to get a review somewhere between these two dates and the goal is to have your PR merged before the second date

**Comment by shoemoney** [comment]
Thanks for the LGTM and for the cycle dates — understood, I'll expect the final review in the 15 Sep–30 Oct window and keep the branch rebased on 10.11 in the meantime.
